Add the ability to provide own scheduled executor for SMTP timeout

egor-ponomarev opened the pull request:
Hello! I would like to propose an improvement for code related to WriteTimeoutSocket (SMTP). Currently for each session a new executor pool is created: Executors.newScheduledThreadPool(1). In my point of view it would be very convenient to be able to specify the executor to not create the new one each time. 
What do you think about it?

I didn't add tests, if the code looks good please let me know and I will add some tests.

jmehrens commented:
@egor-ponomarev Thanks for the contribution!  The tricky part of this is the lifecycle management of the common pool in [WriteTimeoutSocket.close](https://github.com/eclipse-ee4j/angus-mail/blob/2b272d281484b471369622d7abb6d779c39541c0/core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java#L271).

I believe this is why there is that comment of `// XXX - could share executor with all instances?`.  It was easy to tie the lifecycle of the pool to the life of the socket and keep the scope hidden.  I would imagine the patch in its current form would fail to enforce the write timeouts after the first socket was closed.

To move forward we need to define some of the rules for how the pool lifecycle is managed for when the property is not defined and when it is defined.  If the user is supplying an executor then the user should perform the shutdown.  If none is given we need to define when we perform the shutdown.  I think the widest scope we can go is per Service instance so we can tie shutdown to the close of the Service.

We should also consider updating the code to no longer ignore RejectedExecutionException if we are supplying a thread pool from an outside source.  It should be wrapped as an IOException and thrown.  We also need to consider what to do if the external pool invokes shutdownNow and steals the future task.  That case will result in a missed deadline.

We should also benchmark with setRemoveOnCancelPolicy to true as the write methods will create garbage that would only be reclaimed after the timeout.  If the write time out is really long that is a lot of garbage being retained.

egor-ponomarev commented:
@jmehrens  many thanks for your review and provided remarks! I’ve implemented fixes according to them.
Added additional field 'private boolean isExternalSes;’ to know if we work with own scheduled executor service or with an external. External executor is shutdown in close() method. For the internal executor I call setRemoveOnCancelPolicy = true, for external one, I added a comment in docs, recommending users to have true. Would it be ok?

>We also need to consider what to do if the external pool invokes shutdownNow and steals the future task. That case will result in a missed deadline.
Could you please provide any suggestions on what we can do with this?

jmehrens commented on core/src/main/java/com/sun/mail/util/PropUtil.java:
```diff
@@ -149,6 +162,20 @@ private static int getInt(Object value, int def) {
+	 * Interpret the value object as an integer,
+	 * returning def if unable.
+	 */
+	private static ScheduledExecutorService getScheduledExecutorService(Object value) {
```
Copy pasted API docs.  Maybe just delete it?

egor-ponomarev replied:
done, deleted

jmehrens commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -50,16 +50,26 @@ public class WriteTimeoutSocket extends Socket {
     private final ScheduledExecutorService ses;
+    // flag to indicate whether scheduled executor is provided from outside or
+    // should be created here in constructor
+    private boolean isExternalSes;
```
Add final

egor-ponomarev replied:
fixed

jmehrens commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -405,8 +424,9 @@ public synchronized void write(byte[] bs, int off, int len)
-		// ignore it; Executor was shut down by another thread,
-		// the following write should fail with IOException
+            IOException ioex = new IOException("Exception writing in socket");
+            ioex.initCause(ex);
```
Change "Exception writing in socket" to something like "Write aborted due to timeout not enforced."  

We could even query the socket to see if it was closed and provide that information in the message too.  Something to think about since that detail is lost because we no longer allow the attempted write.

egor-ponomarev replied:
message fixed and added status of the socket

jmehrens replied:
A test should be added to show that a shutdown executor with an open socket throws the new exception with the chained rejected execution exception.

jmehrens commented on providers/smtp/src/main/java/com/sun/mail/smtp/package-info.java:
```diff
@@ -226,6 +226,17 @@
+ * <TD>java.util.concurrent.ScheduledExecutorService</TD>
+ * <TD> Provides specific ScheduledExecutorService for mail.smtp.writetimeout option.
+ * The value of mail.smtp.writetimeout shouldn't be a null.
+ * It is highly recommended for provided executor have a setRemoveOnCancelPolicy = true.
```
We should use the javadoc link to the SecudeuledThreadPoolExecutor::setRemoveOnCancelPolicy

egor-ponomarev replied:
fixed

jmehrens reviewed (changes requested):
Looking good. You'll have to do the fun part of adding a unit tests for:

1. PropUtil behavior (null and non null)
2. WriteTimeoutSocket test to prove external executor is being used.
3. WriteTimeoutSocket test to prove WTS instance A close doesn't cause failures instance WTS B using the default executor.
4. WriteTimeoutSocket test to prove WTS instance close doesn't shutdown the external executor.
5. Prove the default executor is shutdown when WTS close.  Maybe just use some reflection to access the private field.

jmehrens commented:
>> We also need to consider what to do if the external pool invokes shutdownNow and steals the future task. That case will result in a missed deadline.
> 
> Could you please provide any suggestions on what we can do with this?

I'm still thinking about this one and are options are not great.  What we would like to have happen is that the socket gets closed if shutdownNow is called.  The only way to do that is to schedule task that just sleeps forever and if it is interrupted close the socket.  I really don't like that solution.  So our other solution is to just live with it and add some documentation to say don't call shutdownNow but rather call shutdown + awaitTermination.

egor-ponomarev commented:
I've just realised that this code not only related to SMTP protocol but IMAP also, I will add changes in docs of IMAP also and check POP3

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/PropUtil.java:
```diff
@@ -149,6 +162,16 @@ private static int getInt(Object value, int def) {
+			return (ScheduledExecutorService) value;
+		}
+
+		return null;
```
@jmehrens btw, wouldn't it be better to throw here an illegal argument exception, what do you think?

jmehrens replied:
ClassCastException would also be fine too.  Failing is probably appropriate here. I forgot that we will have to add a test case for wrong type anyway.

egor-ponomarev replied:
fixed

egor-ponomarev commented:
> Looking good. You'll have to do the fun part of adding a unit tests for:
> 
> 1. PropUtil behavior (null and non null)
> 2. WriteTimeoutSocket test to prove external executor is being used.
> 3. WriteTimeoutSocket test to prove WTS instance A close doesn't cause failures instance WTS B using the default executor.
> 4. WriteTimeoutSocket test to prove WTS instance close doesn't shutdown the external executor.
> 5. Prove the default executor is shutdown when WTS close.  Maybe just use some reflection to access the private field.

@jmehrens could you tell me if I can use Mockito in tests? Or it is not used on purpose?

jmehrens commented:
> > Looking good. You'll have to do the fun part of adding a unit tests for:
> > 
> > 1. PropUtil behavior (null and non null)
> > 2. WriteTimeoutSocket test to prove external executor is being used.
> > 3. WriteTimeoutSocket test to prove WTS instance A close doesn't cause failures instance WTS B using the default executor.
> > 4. WriteTimeoutSocket test to prove WTS instance close doesn't shutdown the external executor.
> > 5. Prove the default executor is shutdown when WTS close.  Maybe just use some reflection to access the private field.
> 
> @jmehrens could you tell me if I can use Mockito in tests? Or it is not used on purpose?

Good question.  @lukasj @jbescos Might be able to answer that.  I would think for this test you could simply create a subclass of OutputStream that just tracks if the close method was called.  For executor testing you can just subclass STPE to track if the shutdown methods were called.

lukasj commented:
> > could you tell me if I can use Mockito in tests? Or it is not used on purpose?
> 
> Good question. @lukasj @jbescos Might be able to answer that.

No, the preference is to avoid dependency on Mockito or any other mocking framework

egor-ponomarev commented:
> Looking good. You'll have to do the fun part of adding a unit tests for:
> 
> 1. PropUtil behavior (null and non null)
> 2. WriteTimeoutSocket test to prove external executor is being used.
> 3. WriteTimeoutSocket test to prove WTS instance A close doesn't cause failures instance WTS B using the default executor.
> 4. WriteTimeoutSocket test to prove WTS instance close doesn't shutdown the external executor.
> 5. Prove the default executor is shutdown when WTS close.  Maybe just use some reflection to access the private field.

Added 1-5 tests, additionally added test for each constructor of WTS checking that only one of them can set external executor

egor-ponomarev commented:
> I've just realised that this code not only related to SMTP protocol but IMAP also, I will add changes in docs of IMAP also and check POP3
fixed

egor-ponomarev commented:
@jmehrens I fixed all points you mentioned, could you please check if everything is ok?

jmehrens commented on core/src/main/java/com/sun/mail/util/PropUtil.java:
```diff
@@ -149,6 +164,16 @@ private static int getInt(Object value, int def) {
+			return (ScheduledExecutorService) value;
+		}
+
+		throw new ClassCastException("Wrong class of scheduled executor");
```
Given the way this method works we could just reduce it to `return (ScheduledExecutorService) value;` or `ScheduledExecutorService.class.cast(value)`.  I've found that the latter is a bit lazier with the class verifier.  Both will have more detailed messages about the types at play versus what we have.

egor-ponomarev replied:
fixed

jmehrens commented on providers/imap/src/main/java/com/sun/mail/imap/package-info.java:
```diff
@@ -333,6 +333,22 @@
+ * {@link java.util.concurrent.ScheduledThreadPoolExecutor#setRemoveOnCancelPolicy(boolean)}.
+ * Without it write methods will create garbage that would only be reclaimed after the timeout.
+ * Do not call {@link java.util.concurrent.ScheduledThreadPoolExecutor#shutdownNow()} in your executor,
+ * it can kill the running task.
```
It would be ok to use shutdownNow when JavaMail sockets are closed.  This would be all service subclasses (Store/Transport).  Invoking run on the returned Future objects would force close the open connections.

We'll have to think of how to capture that idea a bit better.

egor-ponomarev replied:
Fixed with the provided suggestion

jmehrens reviewed (changes requested):
Nice work on the tests.  You should run SpotBugs on all the changes to make sure resources are being closed.  In general Sockets and OutputStreams should be finally closed,  Exeuctors should be finally shutdown.

egor-ponomarev commented:
@jmehrens I was thinking about the way how a service that uses angus-mail library could get information about how many write timeouts happened. Currently, it will get something like SocketClosed, but without any details, if it was a write timeout or something else. In the simplest case, we can add a log in: 

```
	timeoutTask = new Callable<Object>() {
	    @Override
	    public Object call() throws Exception {
		os.close();	// close the stream to abort the write
               log.warn("Write timeout etc");
		return null;
	    }
	};
```
But ideally, it would be better to return to a user the reason of failed sending. In that case, the user will be able to collect some metrics. What do you think?

jmehrens commented:
> @jmehrens I was thinking about the way how a service that uses angus-mail library could get information about how many write timeouts happened. Currently, it will get something like SocketClosed, but without any details, if it was a write timeout or something else. In the simplest case, we can add a log in:
> 
> ```
> 	timeoutTask = new Callable<Object>() {
> 	    @Override
> 	    public Object call() throws Exception {
> 		os.close();	// close the stream to abort the write
>                log.warn("Write timeout etc");
> 		return null;
> 	    }
> 	};
> ```
> 
> But ideally, it would be better to return to a user the reason of failed sending. In that case, the user will be able to collect some metrics. What do you think?

Doesn't look like there is a logger in this class so that can be tricky to add one. 

What we could do is change the Callable to return a String, write timeout exceeded or the message from IOException produced by close.  Then in the TimeoutOutputStream we can hold on to the last scheduled future which will provide us with the meaningful message.  We could then extract and chain that message as an IOException in the TimeoutOutputStream and throw it to the caller. So the output would look like `java.io.IOException: Socket closed caused by java.io.IOException write timeout exceeded.`

egor-ponomarev commented:
> > @jmehrens I was thinking about the way how a service that uses angus-mail library could get information about how many write timeouts happened. Currently, it will get something like SocketClosed, but without any details, if it was a write timeout or something else. In the simplest case, we can add a log in:
> > ```
> > 	timeoutTask = new Callable<Object>() {
> > 	    @Override
> > 	    public Object call() throws Exception {
> > 		os.close();	// close the stream to abort the write
> >                log.warn("Write timeout etc");
> > 		return null;
> > 	    }
> > 	};
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But ideally, it would be better to return to a user the reason of failed sending. In that case, the user will be able to collect some metrics. What do you think?
> 
> Doesn't look like there is a logger in this class so that can be tricky to add one.
> 
> What we could do is change the Callable to return a String, write timeout exceeded or the message from IOException produced by close. Then in the TimeoutOutputStream we can hold on to the last scheduled future which will provide us with the meaningful message. We could then extract and chain that message as an IOException in the TimeoutOutputStream and throw it to the caller. So the output would look like `java.io.IOException: Socket closed caused by java.io.IOException write timeout exceeded.`

Honestly, I didn't get completely the idea:
Do you mean that we need to surround `os.write(bs, off, len);` and then somehow get result of callable method? 
I prepared an example with try catch around it and in catch I check the future was completed:
https://gist.github.com/egor-ponomarev/89ab5d0c6bab3a128603fd39aff11665
But sometimes it will give false positive

egor-ponomarev commented:
> Nice work on the tests. You should run SpotBugs on all the changes to make sure resources are being closed. In general Sockets and OutputStreams should be finally closed, Exeuctors should be finally shutdown.

Did you mean to fix closing resources in tests, didn't you? Fixed. I ran SpotBugs and looks there isn't any issue with my code, there are 2 bugs but they are in 2 old other classes

jmehrens commented:
>Honestly, I didn't get completely the idea:
>Do you mean that we need to surround os.write(bs, off, len); and then somehow get result of callable method?
I'll have to create a branch and write it up but the general idea is:

public String call() {
        try {
		os.close();	// close the stream to abort the write
         } catch(Throwable t) {
            return t.toString();
         }
	return "Write timed out";
    }

With that signature the scheduled Future<String> will either be canceled when the write works or Future::get will return the reason the string is failing.

egor-ponomarev commented:
> > Honestly, I didn't get completely the idea:
> > Do you mean that we need to surround os.write(bs, off, len); and then somehow get result of callable method?
> > I'll have to create a branch and write it up but the general idea is:
> 
> public String call() { try { os.close(); // close the stream to abort the write } catch(Throwable t) { return t.toString(); } return "Write timed out"; }
> 
> With that signature the scheduled Future will either be canceled when the write works or Future::get will return the string reason for failing.
> 
> } catch(IOException e) { if (sf.isDone()) { throw new IOException(sf.get(), e); //get throws checked and unchecked exceptions. They all need to get wrapped as IOException } else { throw e; }

Thanks @jmehrens , I think I get it, I will implement

egor-ponomarev commented:
> > > Honestly, I didn't get completely the idea:
> > > Do you mean that we need to surround os.write(bs, off, len); and then somehow get result of callable method?
> > > I'll have to create a branch and write it up but the general idea is:
> > 
> > 
> > public String call() { try { os.close(); // close the stream to abort the write } catch(Throwable t) { return t.toString(); } return "Write timed out"; }
> > With that signature the scheduled Future will either be canceled when the write works or Future::get will return the string reason for failing.
> > } catch(IOException e) { if (sf.isDone()) { throw new IOException(sf.get(), e); //get throws checked and unchecked exceptions. They all need to get wrapped as IOException } else { throw e; }
> 
> Thanks @jmehrens , I think I get it, I will implement

@jmehrens https://github.com/eclipse-ee4j/angus-mail/pull/22/commits/72f285927a5be4b9f14bf81787abe178da12338c
Would it be ok like this?
I wanted to add tests for TimeoutOutputStream but since this class is nested with default access it is not a trivial task. Maybe we can mark it as public or make it a normal class, not the nested one, what do you think?

jmehrens commented:
> @jmehrens [72f2859](https://github.com/eclipse-ee4j/angus-mail/commit/72f285927a5be4b9f14bf81787abe178da12338c) Would it be ok like this? I wanted to add tests for TimeoutOutputStream but since this class is nested with default access it is not a trivial task. Maybe we can mark it as public or make it a normal class, not the nested one, what do you think?

For these cases I just create a private helper method in the JUnit test that just resorts to reflection.  You should be able to use Class.forName or WriteTimeoutSocket.class.getDeclaredClasses to access the TimeoutOutputStream class.  Then set it accessible and invoke the declared constructor.

egor-ponomarev commented:
> > @jmehrens [72f2859](https://github.com/eclipse-ee4j/angus-mail/commit/72f285927a5be4b9f14bf81787abe178da12338c) Would it be ok like this? I wanted to add tests for TimeoutOutputStream but since this class is nested with default access it is not a trivial task. Maybe we can mark it as public or make it a normal class, not the nested one, what do you think?
> 
> For these cases I just create a private helper method in the JUnit test that just resorts to reflection. You should be able to use Class.forName or WriteTimeoutSocket.class.getDeclaredClasses to access the TimeoutOutputStream class. Then set it accessible and invoke the declared constructor.

sorry, my bad, noticed it is not a nested class, just a class with default access located below the public class in the same file

egor-ponomarev commented:
I've added tests for TimeoutOutputStream class. 
@jmehrens , I've finished the fixes, could you please check if everything is fine?
We've had several conversations, maybe I forgot something, if so, please let me know

jmehrens commented on providers/angus-mail/src/test/java/com/sun/mail/util/PropUtilTest.java:
```diff
@@ -179,4 +185,30 @@ public void testSystemBoolean() throws Exception {
+    @Test
+    public void testScheduledExecutorWriteTimeout() {
+        final String executorPropertyName = "test";
+        ScheduledExecutorService ses = new ScheduledThreadPoolExecutor(1);
```
We should shutdown the ses after we create it just to be future proof

egor-ponomarev replied:
fixed

jmehrens commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+
+		timeoutOutputStream.write(new byte[]{0}, 0, 1);
+
+		Assert.assertEquals(0, queue.size());
```
Finally shutdownNow the executor as caller is managing it.

jmehrens commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+
+		Assert.assertEquals(1, queue.size());
+		ScheduledFuture<String> sf = (ScheduledFuture<String>) queue.peek();
+		Assert.assertTrue(sf.isCancelled());
```
Finally shutdown ses

jmehrens commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+			() -> timeoutOutputStream.write(new byte[] {1}, 0, 1));
+
+		Assert.assertEquals("java.lang.RuntimeException: Unknown error", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
```
Finally shutdownNow ses and finally close socket

egor-ponomarev replied:
fixed

jmehrens reviewed (changes requested):
Minor clean ups.  I think in general we should be shutting down the `ses` before test methods end. and closing client sockets before test methods end.

egor-ponomarev commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+				try {
+					Thread.sleep(delay.toMillis());
+				} catch (InterruptedException e) {
+					e.printStackTrace();
```
fixed

egor-ponomarev commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+			() -> timeoutOutputStream.write(new byte[] {1}, 0, 1));
+
+		Assert.assertEquals("Write timed out", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
```
fixed

egor-ponomarev commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+
+		timeoutOutputStream.write(new byte[]{0}, 0, 1);
+
+		Assert.assertFalse(socket.isClosed());
```
fixed

egor-ponomarev commented on providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java:
```diff
@@ -0,0 +1,223 @@
+			Assert.assertThrows(IOException.class, () -> timeoutOutputStream.write(new byte[] {0}, 0, 1));
+
+		Assert.assertEquals("java.net.SocketException: Socket closed", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
```
fixed

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -419,4 +457,12 @@ public synchronized void write(byte[] bs, int off, int len)
+        } catch (Exception e) {
+            return String.format("%s. %s", CANNOT_GET_TIMEOUT_TASK_RESULT_MESSAGE, e.getMessage());
+        }
+    }
```
fixed

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -158,7 +169,7 @@ public InputStream getInputStream() throws IOException {
 	// wrap the returned stream to implement write timeout
-        return new TimeoutOutputStream(socket.getOutputStream(), ses, timeout);
+        return new TimeoutOutputStream(socket.getOutputStream(), ses, timeout, socket);
     }
```
good point, fixed

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -398,17 +427,26 @@ public synchronized void write(byte[] bs, int off, int len)
+            os.write(bs, off, len);
+        } catch (Exception e) {
+            IOException exception = sf == null ? new IOException(e) : new IOException(getTimeoutTaskResult(sf), e);
+            throw exception;
```
I was thinking, isn't it enough just to catch IOException? Or do you think it is better to catch RuntimeException also?

egor-ponomarev replied:
I did change it to catch only IOException but can add RuntimeException if needed

jmehrens replied:
Yes you are right. IOException is all that is needed.

jmehrens reviewed (changes requested):
Good job! We are getting this down to just little minor issues and edge cases.

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -419,4 +456,20 @@ public synchronized void write(byte[] bs, int off, int len)
+            if (e instanceof InterruptedException) {
+                wasInterrupted = true;
+            }
+            return String.format("%s. %s", CANNOT_GET_TIMEOUT_TASK_RESULT_MESSAGE, e.getMessage());
```
@jmehrens thanks again for the review! No problem, I'll change it. However, I'm not sure if I got properly your idea. If we have 3 catch blocks, we will have something like:
```
    private String getTimeoutTaskResult(ScheduledFuture<String> sf)  {
        boolean wasInterrupted = Thread.interrupted();
        String exceptionMessage = null;
        try {
            return sf.get();
        } catch (CancellationException e) {
            exceptionMessage = e.toString();
        } catch (InterruptedException e) {
            exceptionMessage = e.toString();
            wasInterrupted = true;
        } catch (ExecutionException e) {
            exceptionMessage = e.toString();
        } catch (Exception e) {
            exceptionMessage = e.toString();
        } finally {
            if (wasInterrupted) {
                Thread.currentThread().interrupt();
            }
        }

        return String.format("%s. %s", CANNOT_GET_TIMEOUT_TASK_RESULT_MESSAGE, exceptionMessage);
    }
```
But what is the difference if eventually, we have e.toString() for all cases except InterruptedException (where we set a boolean variable)? Also, I think we should keep Exception, just in case something goes wrong it won't impact the execution, what do you think?

egor-ponomarev replied:
fixed

jmehrens commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -398,17 +426,26 @@ public synchronized void write(byte[] bs, int off, int len)
+        }
 	} finally {
 	    if (sf != null)
 		sf.cancel(true);
```
I just noticed bug here.  We have to make sure that cancel is called before we get the result or we are going to end up blocking for full timeout every call.

We have one of two options here:

1.  Add another try block so the finally block can run before we catch IOException.  Then getTimeoutTaskResult should add an if condition to not call sf.get when sf.isCancelled() is true.
2.  Call sf.cancel(true) in getTimeoutTaskResult and if the result is false proceed with sf.get.

egor-ponomarev replied:
fixed using the 2nd suggestion

egor-ponomarev replied:
I will also try to add a test for checking exceptions caught in handleTimeoutTaskResult.
Btw, do you mind if I squash all commits of the feature into one?

jmehrens replied:
Squashing the commits in this branch is fine.

jmehrens commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -398,25 +428,57 @@ public synchronized void write(byte[] bs, int off, int len)
 
     @Override
     public void close() throws IOException {
 	os.close();
```
Think this should be:
 os.close();
  if (sf != null)
    sf.cancel(true);
 }

Thoughts?

egor-ponomarev replied:
good point, looks like we should make sf a member variable so that it would be accessible by close method

egor-ponomarev replied:
fixed

jmehrens reviewed (changes requested):
Thanks for working hard on this!  I need to take a look at the test some so more to come.

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -398,25 +428,57 @@ public synchronized void write(byte[] bs, int off, int len)
+        try {
+            if (!sf.cancel(true)) {
+                return sf.get(0, TimeUnit.NANOSECONDS);
+            }
```
@jmehrens thank you for your review! I'm on vacation this week and will apply the suggested changes next week as soon as I'm back

egor-ponomarev replied:
Hello @jmehrens , I'm back, sorry again for the delay. Your points make sense, just one question about the last one, do you suggest having something like that (not pass original IOException when sf.cancel is true):

```       
        try {
            os.write(bs, off, len);
        } catch (IOException e) {
            if (sf != null && sf.cancel(true)) {
                throw new IOException("The write wasn't completed due to write timeout");
            }
            e = sf == null ? e : new IOException(handleTimeoutTaskResult(sf), e);
            throw e;
        }
```
?

jmehrens replied:
Hello!  No worries on the delay.  I think what you have is correct.  I think we can get rid of some null checking if we write it like:
```
try {
            os.write(bs, off, len);
        } catch (IOException e) {
            if (sf != null && !sf.cancel(true)) {
                throw new IOException(handleTimeoutTaskResult(sf), e)
            }
            throw e;
        }
```

If we are able to

egor-ponomarev replied:
but in this case, the caller will see a message like "Socket closed" without being aware that the write timeout occurred. 
Maybe it would be useful to provide somehow this info, what do you think?

jmehrens replied:
My understanding is: If `sf` is null there is no detailed result to fetch so throw original exception.  If we are able to cancel `sf` it will never run therefore therefore we are under the timeout threshold and there is no detailed result to fetch.  In that case we just throw the original exception.

egor-ponomarev replied:
ah, right, my bad - got confused, thanks

egor-ponomarev replied:
fixed

jmehrens reviewed (changes requested):
Nice work on this!  I know it has been a long road.  Currently I only see one issue which I put in another comment.  So that means I would like to pause for a bit on my review and come back to this code with fresh eyes.  This will help me catch anything I missed.  After than we'll look at next steps.  Thanks for your hard work!

egor-ponomarev commented:
> Nice work on this! I know it has been a long road. Currently I only see one issue which I put in another comment. So that means I would like to pause for a bit on my review and come back to this code with fresh eyes. This will help me catch anything I missed. After than we'll look at next steps. Thanks for your hard work!

@jmehrens many thanks to you for the detailed code reviews and for noticing important points to fix and change! A pause it is not a problem for me, take your time :+1:

egor-ponomarev commented on core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java:
```diff
@@ -454,16 +457,22 @@ public synchronized void write(byte[] bs, int off, int len)
         boolean wasInterrupted = Thread.interrupted();
         String exceptionMessage = null;
         try {
             if (!sf.cancel(true)) {
```
fixed

egor-ponomarev commented:
Also, I've squashed commits into the one so that it would be easier to review changes and have clearer history

jmehrens commented:
@egor-ponomarev I going to try to get this reviewed by the end of the week.  I haven't forgot about it just been delayed.

egor-ponomarev commented:
@jmehrens no problem, thank you in advance!

jmehrens commented:
@egor-ponomarev I did one more review and I can't seem to spot anything major on this.  Did you sign an Eclipse committer agreement?

@lukasj I've reviewed this pull request but would like your review on it too.  I can answer questions on the approach taken.

lukasj reviewed (approved):
LGTM
